### PR TITLE
Reduce duplication

### PR DESCRIPTION
**Pull request by ericpoe**
This is a code-cleanup PR.

This PR reduces a bunch of duplication in the Functional classes.

HStringContains and HArrayMap have been cleaned up so that the intention is easier to discern.

### CODE REVIEW

File: src/Container/HStringContains.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<?php
```
Comment (ericpoe) [inline src/Container/HStringContains.php:1]: I don't know why git thinks this is created here and deleted later when this was just a simple rename.

There are also a few changes in the `contains` method that simplifies its behavior.

File: src/Container/HStringContains.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    public function contains($value)
+    {
+        if ($value instanceof HString) {
+            $value = $value->toString();
+        }
+
+        if (is_scalar($value)) {
+            $this->value = (string) $value;
+        } else {
+            throw new \InvalidArgumentException(sprintf("%s is neither a scalar value nor an HString", Helper::getType($value)));
```
Comment (johnkary) [inline src/Container/HStringContains.php]: [NB] This error message says _what_ is wrong but doesn't describe _why_ it's wrong. What's wrong with passing in a value that's not a scalar or HString? What can't be done that was intended to be done when the developer called this method?

File: src/Container/HStringContains.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    /**
+     * @return bool
+     */
+    private function containsValue()
+    {
+        $pos = strpos($this->str, $this->value);
+
+        return (false !== $pos) ?: false;
```
Comment (johnkary) [inline src/Container/HStringContains.php]: [NB] I find the ternary operator use here makes this confusing. It's easy to reason about but I have to think a minute to be sure I've got it straight.

The true-condition block doesn't explicitly list a value `true`, yet has `false` in the false-condition block. The ternary isn't needed in this case.

That slight quick is compounded by the double-negative `false !==` part of the comparison. `false` ends up appearing twice, and the not-equals can trip me up. So here's my thought process when looking at this function...

"So when false is not equal to $pos... What's the $pos again?
Oh yeah the numeric position if the string is found... unless it's not found, then it's `false`, right?
OK then which part of the ternary operator are we returning...? The missing true-condition? Or wait, the false-condition? And the false-condition is `false` but wait what's the true-condition? A number from strpos() or false?
Or?
😣 
Sigh, I suck at programming."

I'd like using an instance variable to make the return statement easier to read if we were doing more complex true/false work but in the ternary operator, but in this case I think all these issues are solved by simply returning the result of comparing the strpos return value:

```
private function containsValue()
{
    return false !== strpos($this->str, $this->value);
}
```

What do you think?

**Comment by johnkary** [comment]
All this cleanup looks great! `HStringContains` looks much more simple.

A lot of classes changed from "HaystackString*" to "HString*", also "HaystackArray*" to "HArray*". What about the remaining classes that have "Haystack" still in the title? Could these be shortened similarly?
- src/Functional/HaystackFilter.php
- src/Functional/HaystackFilterWithDefaults.php
- src/Functional/HaystackFilterWithKey.php
- src/Functional/HaystackFilterWithValue.php
- src/Functional/HaystackFilterWithValueAndKey.php
- src/Functional/HaystackMap.php
- src/Functional/HaystackReduce.php

File: src/Container/HStringContains.php (hunk 3 above)
Comment (ericpoe) [inline src/Container/HStringContains.php]: That certainly makes more sense. I was struggling to make the line more clear and ended up making it more convoluted!

File: src/Container/HStringContains.php (hunk 2 above)
Comment (ericpoe) [inline src/Container/HStringContains.php]: It's a good thought. I'll think about an appropriate exception text.

**Pull request merged**